cmd/kubeadm/app/discovery/token/token.go

```diff
@@ -181,42 +181,40 @@ func buildSecureBootstrapKubeConfig(endpoint string, caCert []byte, clustername
 	return bootstrapConfig
 }
 
-// runForEndpointsAndReturnFirst loops the endpoints slice and let's the endpoints race for connecting to the master
-func runForEndpointsAndReturnFirst(endpoints []string, discoveryTimeout time.Duration, fetchKubeConfigFunc func(string) (*clientcmdapi.Config, error)) (*clientcmdapi.Config, error) {
-	stopChan := make(chan struct{})
+// fetchKubeConfigWithTimeout tries to run fetchKubeConfigFunc on every DiscoveryRetryInterval, but until discoveryTimeout is reached
+func fetchKubeConfigWithTimeout(apiEndpoint string, discoveryTimeout time.Duration, fetchKubeConfigFunc func(string) (*clientcmdapi.Config, error)) (*clientcmdapi.Config, error) {
+	cancelChan := make(chan struct{})
+	finishedChan := make(chan struct{})
 	var resultingKubeConfig *clientcmdapi.Config
 	var once sync.Once
-	var wg sync.WaitGroup
-	for _, endpoint := range endpoints {
-		wg.Add(1)
-		go func(apiEndpoint string) {
-			defer wg.Done()
-			wait.Until(func() {
-				fmt.Printf("[discovery] Trying to connect to API Server %q\n", apiEndpoint)
-				cfg, err := fetchKubeConfigFunc(apiEndpoint)
-				if err != nil {
-					fmt.Printf("[discovery] Failed to connect to API Server %q: %v\n", apiEndpoint, err)
-					return
-				}
-				fmt.Printf("[discovery] Successfully established connection with API Server %q\n", apiEndpoint)
-
-				// connection established, stop all wait threads
-				once.Do(func() {
-					close(stopChan)
-					resultingKubeConfig = cfg
-				})
-			}, constants.DiscoveryRetryInterval, stopChan)
-		}(endpoint)
-	}
+
+	go func() {
+		defer close(finishedChan)
+		wait.Until(func() {
+			fmt.Printf("[discovery] Trying to connect to API Server %q\n", apiEndpoint)
+			cfg, err := fetchKubeConfigFunc(apiEndpoint)
+			if err != nil {
+				fmt.Printf("[discovery] Failed to connect to API Server %q: %v\n", apiEndpoint, err)
+				return
+			}
+			fmt.Printf("[discovery] Successfully established connection with API Server %q\n", apiEndpoint)
+			once.Do(func() {
+				resultingKubeConfig = cfg
+				close(cancelChan)
+			})
+		}, constants.DiscoveryRetryInterval, cancelChan)
+	}()
+
 	select {
 	case <-time.After(discoveryTimeout):
-		close(stopChan)
+		once.Do(func() {
+			close(cancelChan)
+		})
 		err := fmt.Errorf("abort connecting to API servers after timeout of %v", discoveryTimeout)
 		fmt.Printf("[discovery] %v\n", err)
-		wg.Wait()
+		<-finishedChan
```

Comment:
I think the wait/waitgroup is a lot cleaner here imo.  

The double use of the channel for both the case statement and within the case is an anti-pattern.